docs/wslgsu.1
.TH "WSLGSU" "1" "DATEPLACEHOLDER" "VERSIONPLACEHOLDER" "WSL Utilities User Manual"
.SH NAME
.B wslgsu
- Component of WSL Utilities
.SH SYNOSIS
.B wslgsu
.RB [ \-u
.IR USERNAME ]
.RB [ \-n 
.IR NAME ]
.RB [ \-S ]
.PP
.B wslgsu
.RB [ \-hv ]
.I SERVICE/COMMAND
.SH DESCRIPTION
This tool allows you to generate a WSL startup Task using the Windows Task Scheduler.
.SH OPTIONS
.TP
.B -h, --help
Print a simple help.
.TP
.B -v, --version
Print current version.
.TP
.B -u, --username \fIUSERNAME\fR
Use a separate user to execute the command/service. By default it's \fIroot\fR.
.TP
.B -n, --name \fINAME\fR
Add an explicit name instead of the default of automatically extracting a name from the service/command.
.TP
.B -S, --service
Using this option means you are addng a system service. without \fB-S\fR, you are adding a command to system. 
.SH AUTHOR
Created by Patrick Wu <me@patrickwu.space>
.SH REPORTING BUGS
Report bugs to <https://github.com/wslutilities/wslu/issues>
.SH COPYRIGHT
This is free software; you can redistribute it and/or modify it under
the terms of the GNU GPL version 3 or (at your option) any later
version.
There is NO warranty; not even MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
.SH SEE ALSO
wslu(7)


debian/changelog
wslu (3.2.3-0kali1) kali-dev; urgency=medium

  * wslsys: fix an issue in call_windows_uptime
    (GH: #182)
  * header: fix a distro detection issue in 
    Ubuntu Impish (21.10)
  * debian/control: fix dependency issue in 
    installing wslu (LP: #1926771)
  * docs: include available commands in manpage 
    (GH: #169)
  * project: fix a problem that Powershell output
    non-unicode character (GH: #162)
  * wslview: fix an issue that caused the relative
    Linux path unable to be opened (GH: #165)
  * wslusc: fix an issue that some people unable
    to open their GUI application (GH: #174)
  * header: Avoid error message when the file doesn't
    exist (LP: 1911379)

 -- Patrick Wu <me@patrickwu.space>  Mon, 14 Jun 2021 22:14:49 +0800

wslu (3.2.2-0kali1) kali-dev; urgency=medium

  * docs: include available commands in manpage 
    (GH: #169)
  * project: fix a problem that Powershell output
    non-unicode character (GH: #162)
  * wslview: fix an issue that caused the relative
    Linux path unable to be opened (GH: #165)
  * wslusc: fix an issue that some people unable
    to open their GUI application (GH: #174)

 -- Patrick Wu <me@patrickwu.space>  Mon, 15 Mar 2021 10:52:33 +0800

wslu (3.2.1-0kali1) kali-dev; urgency=medium

  * header: improvements on the system drive detection. (closes #133)
  * tests: update the test system and tests. (closes #125)
  * wslact: add back-compatibility for the auto-mounting feature.
  * wslview: fix an issue when handling http/https
  * wslview: fix a bug that it won't open file:///mnt/c/... or /mnt/c/...
  * wslusc: fix an issue that wsl.exe execute with wrongly passed param (close #138)
  * wslusc: fix the problem that the previous version that GUI shortcut generated won't work.
  * wslsys: fix an issue that the WSL Version shown is wrong.
  * Includes a wslview.desktop (PR #143 and #146)
  * Update the script to fix badly generated shortcut issue (fixes #141)
  * Fix a DPI script related space issue (fixes #144)

 -- Patrick Wu <me@patrickwu.space>  Wed, 23 Dec 2020 22:33:07 +0800

wslu (3.1.1-0kali1) kali-dev; urgency=medium

  * project: add separate Fedora Remix support
  * header: fix `ubuntu.exe` detection issue
  * wslview: mistake in calling a variable (fixes #118 again)
  * wslusc: always use the absolute path (fixes #129)
  * wslusc: fixes a possible white space path issue
  * wslusc: When TMP is empty, try TEMP

 -- Patrick Wu <me@patrickwu.space>  Wed, 01 Jul 2020 22:22:53 +0800

wslu (3.1.0-0kali1) kali-dev; urgency=medium

  * header: use the proper color definition
  * wslfetch: new display for WSL with WSL version displayed
  * wslfetch: use the proper logo for Ubuntu (Reported by @flexiondotorg)
  * wslsys: add WSL version option
  * wslsys: fix an issue in DPI scaling getting wrong value due to the API change in Latest fast ring insider build (Reported by @crramirez)
  * wslact: adding `--mount-options` for auto-mounting feature
  * wslact: fix an i18n related issue in time-sync feature (fixes #113 by @ricardosantos9521)
  * wslgsu: removed temporarily
  * wslgsu: limit the Windows version it can use
  * wslview: properly detect all different protocols and path styles (fixes #118)
  * package: fix alpine dependency issue (fixes #117)

 -- Patrick Wu <me@patrickwu.space>  Wed, 13 May 2020 22:53:03 +0800

wslu (3.0.1-0kali1) kali-dev; urgency=medium

  * Initial release

 -- Patrick Wu <me@patrickwu.space>  Mon, 13 Apr 2020 20:41:40 +0800


src/wslgsu.sh
# shellcheck shell=bash
isService=0
isWakeup=0
wa_gs_commd=""
wa_gs_dscp=""
wa_gs_name=""
wa_gs_user="root"

help_short="wslgsu [-u USERNAME] [-n NAME] [-S] SERVICE/COMMAND\nwslgsu [-hvw]"

while [ "$1" != "" ]; do
	case "$1" in
		-w|--wakeup) isWakeup=1; break;;
		-u|--user) shift; wa_gs_user="$1"; shift;;
		-n|--name) shift; wa_gs_name="$1"; shift;;
		-S|--service) isService=1; shift;;
		-h|--help) help "wslact" "$help_short"; exit;;
		-v|--version) version; exit;;
		*) wa_gs_commd="$*";break;;
	esac
done

debug_echo "isService: $isService"
debug_echo "isWakeup: $isWakeup"
debug_echo "wa_gs_commd: $wa_gs_commd"
debug_echo "wa_gs_dscp: $wa_gs_dscp"
debug_echo "wa_gs_name: $wa_gs_name"

wslutmpbuild=$(wslu_get_build)
[ "$wslutmpbuild" -ge "$BN_MAY_NINETEEN" ] || (echo "This tool is not supported before version 1903."; exit 34)

if [[ "$wa_gs_commd" != "" ]] || [[ $isWakeup -eq 1 ]]; then
	debug_echo "command or wakeup exist, executing"
	tmp_location="$(wslvar -s TMP)"
	tpath="$(double_dash_p "$tmp_location")" # Windows Temp, Win Double Sty.
	tpath_linux="$(wslpath "$tmp_location")" # Windows Temp, Linux WSL Sty.
	script_location_win="$(wslvar -s USERPROFILE)\\wslu" #  Windows wslu, Win Double Sty.
	script_location="$(wslpath "$script_location_win")" # Windows wslu, Linux WSL Sty.

	debug_echo "tmp_location: $tmp_location"
	debug_echo "tpath: $tpath"
	debug_echo "tpath_linux: $tpath_linux"
	debug_echo "script_location_win: $script_location_win"
	debug_echo "script_location: $script_location"

	# Check presence of sudo.ps1 and 
	wslu_file_check "$script_location" "sudo.ps1"
	wslu_file_check "$script_location" "runHidden.vbs"

	# check if it is a service, a command or it just want to wakeup
	if [[ $isWakeup -eq 1 ]]; then
		debug_echo "Entering wakeup"
		# handling no name given case
		if [[ "$wa_gs_name" = "" ]]; then
			debug_echo "No name given, using default"
			wa_gs_name="Wakeup"
		fi
		wa_gs_commd="wsl.exe -d $WSL_DISTRO_NAME echo"
		wa_gs_dscp="Wake up WSL Distro $WSL_DISTRO_NAME when computer start up; Generated By WSL Utilities"
	elif [[ $isService -eq 1 ]]; then
	# service
		debug_echo "Entering service"
		# handling no name given case
		if [[ "$wa_gs_name" = "" ]]; then
			debug_echo "No name given, using default"
			wa_gs_name="$wa_gs_commd"
		fi
		wa_gs_commd="wsl.exe -d $WSL_DISTRO_NAME -u $wa_gs_user service $wa_gs_commd start"
		wa_gs_dscp="Start service $wa_gs_name from $WSL_DISTRO_NAME when computer start up; Generated By WSL Utilities"
	else
	# command
		debug_echo "Entering command"
		# handling no name given case
		if [[ "$wa_gs_name" = "" ]]; then
			debug_echo "No name given, automatically generate"
			wa_gs_name=$(basename "$(echo "$wa_gs_commd" | awk '{print $1}')")
		fi
		wa_gs_commd="wsl.exe -d $WSL_DISTRO_NAME -u $wa_gs_user $wa_gs_commd"
		wa_gs_dscp="Executing following command \`$wa_gs_name\` from $WSL_DISTRO_NAME when computer start up; Generated By WSL Utilities"
	fi

	debug_echo "wa_gs_commd: $wa_gs_commd"
	debug_echo "wa_gs_dscp: $wa_gs_dscp"

	# Keep the name always unique; such random, much wow
	tmp_rand="$(tr -dc 'a-zA-Z0-9' < /dev/urandom | fold -w 6 | head -n 1)"

	debug_echo "tmp_rand: $tmp_rand"

	# shellcheck disable=SC2028
	tee "$tpath_linux"/tmp.ps1 >/dev/null << EOF 
Import-Module 'C:\\WINDOWS\\system32\\WindowsPowerShell\\v1.0\\Modules\\Microsoft.PowerShell.Utility\\Microsoft.PowerShell.Utility.psd1';
\$action = New-ScheduledTaskAction -Execute 'C:\\Windows\\System32\\wscript.exe' -Argument '$script_location_win\\runHidden.vbs $wa_gs_commd';
\$settings = New-ScheduledTaskSettingsSet -AllowStartIfOnBatteries;
\$trigger =  New-ScheduledTaskTrigger -AtLogOn -User \$env:userdomain\\\$env:username; \$trigger.Delay = 'PT2M';
\$task = New-ScheduledTask -Action \$action -Trigger \$trigger -Description \"$wa_gs_dscp\" -Settings \$settings;
Register-ScheduledTask -InputObject \$task -TaskPath '\\' -TaskName 'WSLUtilities_Actions_Startup_${wa_gs_name}_${tmp_rand}' | out-null;
EOF

	debug_echo "$(cat "$tpath_linux"/tmp.ps1)"
	echo "${warn} WSL Utilities is adding \"${wa_gs_name}\" to Task Scheduler; A UAC Prompt will show up later. Allow it if you know what you are doing."
	if winps_exec "$script_location_win"\\sudo.ps1 "$tpath"\\tmp.ps1; then
		debug_echo "Task Scheduler added"
		rm -rf "$tpath_linux/tmp.ps1"
		echo "${info} Task \"${wa_gs_name}\" added."
	else
		rm -rf "$tpath_linux/tmp.ps1"
		error_echo "Adding Task \"${wa_gs_name}\" failed." 1
	fi
else
	error_echo "No input, aborting" 21
fi

unset isService
unset wa_gs_commd
unset wa_gs_name
unset wa_gs_user
